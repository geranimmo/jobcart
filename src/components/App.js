import React, { Component } from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import reducers from './reducers';
import RouterComponent from './Routers';

export class App extends Component {
	render() {
		const store = createStore(reducers, {}, applyMiddleware(thunk));

		return (
			<Provider store={store}>
				<RouterComponent />
			</Provider>
		);
	}
}

export default App;