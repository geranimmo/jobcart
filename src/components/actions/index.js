export * from './Authentication';
export * from './GetListPackage';
export * from './AddToCart';
export * from './RemoveFromCart';
export * from './GetTotalCost';
export * from './GetTotalDiscount';
