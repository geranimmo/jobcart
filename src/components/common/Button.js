import React from "react";

const Button = props => {
	return <button type="button" {...props}>{props.value}</button>;
};

export { Button };
