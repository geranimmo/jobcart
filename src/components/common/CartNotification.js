import React from "react";

const CartNotification = props => {
	const {
		totalItems,
		className
	} = props;
	const cartTotal = totalItems > 9 ? '9+' : totalItems;
    
	return <div className={`${className} ${totalItems === 0 ? 'hidden' : ''}`}>{cartTotal}</div>;
};

export { CartNotification };
