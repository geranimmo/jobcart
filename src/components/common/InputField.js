import React from "react";

const InputField = props => {
	return <input {...props}/>;
};

export { InputField };
