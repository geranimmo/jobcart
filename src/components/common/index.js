export { InputField } from './InputField';
export { Button } from './Button';
export { CartFooter } from './CartFooter';
export { DateTimeConverter } from './DateTimeConverter';
export { SubstringText } from './SubstringText';
export { CartItem } from './CartItem';
export { CartNotification } from './CartNotification';