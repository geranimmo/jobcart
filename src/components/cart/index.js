import React, { Component } from 'react';
import { withRouter } from "react-router-dom";
import { connect } from 'react-redux';
import _ from 'lodash';
import Header from '../header';
import { CartItem, CartFooter } from '../common';
import { removeFromCart } from '../actions';
import './cart.less';

export class Cart extends Component {
	constructor(props) {
		super(props);
		this.state = {
			myCartList: [],
			totalCartDiscount: 0,
			totalCartCost: 0
		};
	}
    
	componentDidMount() {
		if (!this.props.UserLogin) {
			this.props.history.push('/');
		}
		
		document.addEventListener('scroll', this.handleScroll, true);
	}
    
	componentWillMount() {
		if (this.props.UserLogin) {
			const { ShoppingCart, Packages } = this.props;
			const myCartArr = _.map(ShoppingCart, (item) => {
				return _.merge(item, _.find(Packages, { 'package_id' : item.id }));
			});

			this.setState({ myCartList: myCartArr });

			setTimeout(() => {
				this.populatePromoItems();
				this.populateTotalCost();
			}, 0);
		}
	}
    
    handleScroll = () => {
    	let targetElm = document.getElementById("content__scroller").getBoundingClientRect();
    	let targetOffsetTop = targetElm.top;
    	let header = document.getElementById("header");
		
    	if (targetOffsetTop < 0) {
    		header.classList.add("header__on__scrolled");
    	} else {
    		header.classList.remove("header__on__scrolled");
    	}
    }
	
	removeCartItems = (data) => {
		if (this.props.UserLogin) {
			this.props.removeFromCart(data);

			const updateCartList = this.state.myCartList.filter(e => e.buy_time !== data);

			this.setState({ myCartList: updateCartList });
			
			setTimeout(() => {
				this.populatePromoItems();
				this.populateTotalCost();
			}, 0);
		}
	};

	populateTotalCost() {
		const { myCartList } = this.state;
		let totalCost = 0;
	
		if ( myCartList.length > 0 ) {
			for ( let i = 0; i < myCartList.length; i++ ) {
				totalCost = parseFloat((totalCost + myCartList[i].package_price).toFixed(2));
			}
		}
	
		this.setState({ totalCartCost: totalCost });
	}

	populatePromoItems = () => {
		const { myCartList } = this.state;
		const { client_special } = this.props.Profile;
		let totalDiscount = 0;
	
		if ( client_special.length > 0 ) {
			for ( let i = 0; i < client_special.length; i++ ) {
				let getInCartPromoItems = [];
				
				if ( client_special[i].types === 'free__items' ) {
					// CALCULATE FREE-FOR-SAME-AD
					getInCartPromoItems = filterPromoById(client_special[i].package_id, myCartList);
	
					if (getInCartPromoItems.length > 0) {
						let pricePerItem = getInCartPromoItems[0].package_price;
						let getTotalFreeItems = Math.floor(filterPromoById(client_special[i].package_id, myCartList).length / client_special[i].minimum_items);
						totalDiscount = parseFloat((getTotalFreeItems * pricePerItem).toFixed(2));
					}
				} else if ( client_special[i].types === 'discount__items' ) {
					// CALCULATE DISC-FOR-MIN-BUY
					getInCartPromoItems = filterPromoById(client_special[i].package_id, myCartList);
					
					if (getInCartPromoItems.length >= client_special[i].minimum_items) {
						let discountPrice = client_special[i].discount_price;
						totalDiscount = parseFloat((discountPrice * getInCartPromoItems.length).toFixed(2));
					}
				}
			}
		}
	
		this.setState({ totalCartDiscount: totalDiscount });
	};

	render() {
    	const { myCartList } = this.state;

    	return (
    		<div className={`main__container`}>
    			<Header onCart={true} />
    			<main id={this.props.id} className={`cart__component`}>
    				<div
    					id={`content__scroller`}
    					onScroll={this.handleScroll.bind(this)}
    				>
    					<div className={`cart__empty ${myCartList.length ? 'hidden' : ''}`}>Your Cart Still Empty</div>;
    					{ myCartList.map((item, index) => {
    						return (
								<div key={index}>
									<CartItem
										id={`Cart__${item.id}__${index}__${item.buy_time}`}
										className={`cart__list__wrapper`}
										removeCartItems={this.removeCartItems.bind(this, item.buy_time)}
										item={item}
									/>
								</div>
							);
    					}) }
    				</div>
    			</main>
    			<CartFooter className={`cart__footer`} {...this.state}/>
    		</div>
    	);
	}
}

export const filterPromoById = (id, datas) => {
	return datas.filter(obj => obj.package_id === id);
};

const mapStateToProps = state => {
	return state;
};

export default withRouter(connect(mapStateToProps, { removeFromCart })(Cart));