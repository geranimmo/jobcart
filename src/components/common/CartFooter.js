import React from 'react';
import { Button } from './Button';

const CartFooter = props => {
	let fixedCost = parseFloat((props.totalCartCost - props.totalCartDiscount).toFixed(2));
	
	const buttonProps = {
		className: 'button__checkout',
		value: `Checkout (${props.myCartList.length})`
	};
	
	return (
		<div className={props.className}>
			<div>
				<span>
					<label>Total Cost:</label>
					${fixedCost}
				</span>
				<Button {...buttonProps}/>
			</div>
		</div>
	);
};

export { CartFooter };