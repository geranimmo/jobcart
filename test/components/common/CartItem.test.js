import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { CartItem } from '../../../src/components/common';

describe('>>> C A R T  L I S T  I T E M ---- Test & Snapshot <<<', () => {
	let wrapper;
	const propsValue = {
		id: `Cart__Item__List`,
		className: `cart__list__wrapper`,
		item: {
			id: "Classic",
			package_id: "Classic",
			package_name: "Classic",
			package_recommend: false,
			package_description: "Classic Ad Lorem Ipsum is simply dummy text of the printing and typesetting",
			package_price: 269.99,
			package_image: "/assets/images/employeer-image-sample.jpg",
			buy_time: 1537425742948
		}
	};

	beforeEach(() => {
		wrapper = shallow(<CartItem {...propsValue}/>);
	});
       
	it('+++ Should contains props that assigned +++', () => {
		const targetComponent = wrapper.find(`div.${propsValue.className}`).prop('className');
		expect(targetComponent).toContain(`${propsValue.className}`);
	});
    
	it('+++ Should hide Recommended label if package_recommend is false +++', () => {
		expect(wrapper.find('.info__recommended').prop('className')).toContain('hidden');
	});

	it('+++ Capturing Snapshot of CartItem +++', () => {
		const renderedValue = renderer.create(<CartItem {...propsValue}/>).toJSON();
		expect(renderedValue).toMatchSnapshot();
	});
});