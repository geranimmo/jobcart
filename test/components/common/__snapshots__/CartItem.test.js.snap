// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> C A R T  L I S T  I T E M ---- Test & Snapshot <<< +++ Capturing Snapshot of CartItem +++ 1`] = `
<div
  className="cart__list__wrapper"
  id="Cart__Item__List"
>
  <div
    className="cart__info__wrapper"
  >
    <div
      className="cart__info"
    >
      <h1>
        Classic
        <label
          className="info__recommended hidden"
        >
          RECOMMENDED
        </label>
      </h1>
      <h2>
        Classic Ad Lorem Ipsum is simply dummy text of the printing and typesetting
      </h2>
      <span
        className="info__datetime"
      >
        09/20/2018, 1:42:22 PM
      </span>
      <span
        className="info__price"
      >
        $
        269.99
      </span>
    </div>
  </div>
  <div
    className="cart__delete"
    id="Classic__1537425742948"
  >
    <svg
      clipRule="evenodd"
      fillRule="evenodd"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 24h-14c-1.104 0-2-.896-2-2v-17h-1v-2h6v-1.5c0-.827.673-1.5 1.5-1.5h5c.825 0 1.5.671 1.5 1.5v1.5h6v2h-1v17c0 1.104-.896 2-2 2zm0-19h-14v16.5c0 .276.224.5.5.5h13c.276 0 .5-.224.5-.5v-16.5zm-9 4c0-.552-.448-1-1-1s-1 .448-1 1v9c0 .552.448 1 1 1s1-.448 1-1v-9zm6 0c0-.552-.448-1-1-1s-1 .448-1 1v9c0 .552.448 1 1 1s1-.448 1-1v-9zm-2-7h-4v1h4v-1z"
      />
    </svg>
  </div>
</div>
`;
